osutil/disks/disks_linux.go

```diff
@@ -382,44 +390,67 @@ func (d *disk) FindMatchingPartitionUUID(label string) (string, error) {
 				continue
 			}
 
-			partUUID := udevProps["ID_PART_ENTRY_UUID"]
-			if partUUID == "" {
-				return "", fmt.Errorf("cannot get udev properties for device %s (a partition of %s), missing udev property \"ID_PART_ENTRY_UUID\"", partDev, d.Dev())
+			// we should always have the partition uuid, and we may not have
+			// either the partition label or the filesystem label, on GPT disks
+			// we will always have the partition label, but may not have a
+			// filesystem on the partition, on MBR we will never have a
+			// partition label, but may have a filesystem label. It's unclear if
+			// MBR disks can have partitions without a filesystem label.
+			part.partUUID = udevProps["ID_PART_ENTRY_UUID"]
+			if part.partUUID == "" {
+				return fmt.Errorf("cannot get udev properties for device %s (a partition of %s), missing udev property \"ID_PART_ENTRY_UUID\"", partDev, d.Dev())
 			}
 
-			fsLabelEnc := udevProps["ID_FS_LABEL_ENC"]
-			if fsLabelEnc == "" {
-				// it is valid for there to be a partition without a fs
-				// label - such as the bios-boot partition on amd64 pc
-				// gadget systems
-				// in this case just skip this, since we are only matching
-				// by filesystem labels, obviously we cannot ever match to
-				// a partition which does not have a filesystem
-				continue
-			}
-
-			// TODO: maybe save ID_PART_ENTRY_NAME here too, which is the name
-			//       of the partition. this may be useful if this function gets
-			//       used in the gadget update code
-
-			// we always overwrite the fsLabelEnc with the last one, this
-			// has the result that the last partition with a given
-			// filesystem label will be set/found
-			// this matches what udev does with the symlinks in /dev
-			d.fsLabelToPartUUID[fsLabelEnc] = partUUID
+			// on MBR disks we may not have a partition label, so this may be
+			// the empty string
+			part.partLabel = udevProps["ID_PART_ENTRY_NAME"]
```

Comment:
Looked at #9721, we may want to leave a comment here that ID_PART_ENTRY_NAME is libblkid-like encoded (as compared to PARTNAME which appears plain)